common/transport/amqp/src/amqp.ts

```diff
@@ -90,6 +90,16 @@ export class Amqp {
       }
     }, amqp10.Policy.EventHub));
 
+    const amqpErrorHandler = (err) => {
+      debug('amqp10 client error: ' + err.toString());
+      this._fsm.handle('amqpError', err, () => {
+        if (this._disconnectHandler) {
+          debug('calling upper layer disconnect handler');
+          this._disconnectHandler(err);
```

Comment:
Not your problem, but this bothers me.  The mechanism of setting a handler seems correct, but the idea of a lower level object calling an upper level function smells like an inverted dependency.  OK, it's probably technically not one, but it still smells like it.

Would it be better to emit an event instead?  That feels more node-ish.  Is that more node-ish?  The current code feels like a C developer added this.

Just asking the question for my own education/rumination.

Reply:
you're right, and that's because historically we have this 'disconnect' error that was tacked on top of the normal error-bubbling pipeline that we have to keep carrying around while trying not to "double-bubble-up" errors. this needs cleaning up. right now (for this PR) i'm hesitant to touch something that kinda works though.

---
In reply to: [145961017](https://github.com/Azure/azure-iot-sdk-node/pull/114#discussion_r145961017) [](ancestors = 145961017)